l3kernel/l3basics.dtx

```diff
@@ -3412,11 +3417,53 @@
     \token_if_macro:NTF #1
       {
         \exp_after:wN \@@_prefix_arg_replacement:wN
-          \token_to_meaning:N #1 \s_@@_stop \use_iii:nnn
+          \@@_meaning:N #1 \s_@@_stop \use_iii:nnn
       }
       { \scan_stop: }
   }
 %    \end{macrocode}
+%   LuaMeta\TeX{}'s version of \tn{meaning} doesn't give the same output
+%   as the classical one. So we abstract out the idea at a low level, as
+%   we need the details to be correct here.
+%    \begin{macrocode}
+\cs_if_exist:NF \@@_meaning:N
+  { \cs_new_eq:NN \@@_meaning:N \cs_meaning:N }
+%</code>
+%    \end{macrocode}
+%   This is a modified version of code suggested by Max Chernoff
+%   (\url{https://chat.stackexchange.com/transcript/message/67947633#67947633}).
+%   We aim to provide the classical \TeX{}82 appearance but covering \eTeX{}:
+%   due to changes in the engine, the only useful prefix here is |\protected|.
+%   This does not cover tokens other than control sequences, so we only use
+%   it for the internal needed here.
+%    \begin{macrocode}
+%<*lua>
+if status.luatex_engine == "luametatex" then
+  luacmd('@@_meaning:N', function()
+    local csname = scan_csname(true)
+    local parameters = get_macro(csname, false, true)
+    local body = get_macro(csname)
+    if parameters == body then
+      -- If parameters and body are the same, we assume that there
+      -- are no parameters. (Probably a LuaMetaTeX bug...)
+      parameters = ""
+    end
+    local cstoken = token_create(csname)
+    local prefixes = {}
+    if token["getprotected"](cstoken) then
+      prefixes[#prefixes + 1] = "\\protected"
+    end
+    local prefix = table_concat(prefixes, " ")
+    if prefix ~= "" then
+      prefix = prefix .. " "
+    end
+    sprint(-2, string.formatters["%smacro:%s->%s"](prefix, parameters, body))
```

Comment:
The only point to using `string.formatters` is that you can separate the format string definition from its use; if you're not going to store it in a separate variable, it's better to do

```suggestion
    sprint(-2, string.format("%smacro:%s->%s", prefix, parameters, body))
```

or

```suggestion
    sprint(-2, ("%smacro:%s->%s"):format(prefix, parameters, body))
```

Reply:
I find the entire 'fill in the blanks' approach weird, so I just stuck to what you wrote :) I'll use your suggestion.